auction-house/program/src/lib.rs

```diff
@@ -559,8 +559,132 @@ pub mod auction_house {
+            authority,
+            auction_house,
+            wallet_to_use.to_account_info(),
+            auction_house_fee_account.to_account_info(),
+            &auction_house_seeds,
+        )?;
+
+        let purchase_receipt_info = purchase_receipt.to_account_info();
+        let seller_trade_state_key = seller_trade_state.key();
+        let buyer_trade_state_key = buyer_trade_state.key();
+
+        if purchase_receipt_info.data_is_empty() {
+            let purchase_receipt_seeds = [
+                PURCHASE_RECEIPT_PREFIX.as_bytes(),
+                seller_trade_state_key.as_ref(),
+                buyer_trade_state_key.as_ref(),
+                &[purchase_receipt_bump],
+            ];
+
+            create_or_allocate_account_raw(
+                *ctx.program_id,
+                &purchase_receipt_info,
+                &rent.to_account_info(),
+                &system_program,
+                &fee_payer,
+                PURCHASE_SIZE,
+                fee_payer_seeds,
+                &purchase_receipt_seeds,
+            )?;
+        }
+
+        let purchase = Purchase {
+            buyer: buyer.key(),
+            seller: seller.key(),
+            token_account: token_account.key(),
+            auction_house: auction_house.key(),
+            token_mint: token_mint.key(),
+            price: buyer_price,
+            token_size,
+        };
+
+        purchase.try_serialize(&mut *purchase_receipt_info.data.borrow_mut())?;
+
+        let mut accounts = ExecuteSale {
+            buyer: buyer.clone(),
+            seller: seller.clone(),
+            auction_house: auction_house.clone(),
+            token_account: token_account.clone(),
+            token_mint: token_mint.clone(),
+            treasury_mint: treasury_mint.clone(),
+            metadata: metadata.clone(),
+            authority: authority.clone(),
+            seller_trade_state: seller_trade_state.clone(),
+            buyer_trade_state: buyer_trade_state.clone(),
+            free_trade_state: free_trade_state.clone(),
+            seller_payment_receipt_account: seller_payment_receipt_account.clone(),
+            buyer_receipt_token_account: buyer_receipt_token_account.clone(),
+            escrow_payment_account: escrow_payment_account.clone(),
+            auction_house_fee_account: auction_house_fee_account.clone(),
+            auction_house_treasury: auction_house_treasury.clone(),
+            program_as_signer: program_as_signer.clone(),
+            token_program: token_program.clone(),
+            system_program: system_program.clone(),
+            ata_program: ata_program.clone(),
+            rent: rent.clone(),
+        };
+
+        let context = Context::new(ctx.program_id, &mut accounts, ctx.remaining_accounts);
+
+        execute_sale(
+            context,
+            escrow_payment_bump,
+            _free_trade_state_bump,
+            program_as_signer_bump,
+            buyer_price,
+            token_size,
+        )?;
```

Comment:
The program throughs a balance mismatch if execute_sale is performed before the purchase account is allocated because execute_sale moves lamports around manually which conflicts with token transfer done by ceate_or_allocate_raw. This should be fine as any failure will roll back any account changes done by the transaction.

https://forums.solana.com/t/how-do-i-transfer-lamports-between-3-accounts-owned-by-my-program-and-system-program/4294